File: src/main/java/me/nasukhov/bot/command/AskQuestion.java
package me.nasukhov.bot.command;

import me.nasukhov.bot.bridge.IOResolver;
import me.nasukhov.bot.io.Channel;
import me.nasukhov.bot.io.ChannelRepository;
import me.nasukhov.bot.io.Input;
import me.nasukhov.bot.io.Output;
import me.nasukhov.study.ChannelQuestion;
import me.nasukhov.study.ProgressRepository;
import me.nasukhov.study.Time;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

public class AskQuestion implements Handler {
    private static final String NO_MORE_QUESTIONS_LEFT = "У нас пока нет новых вопросов. Проверьте позже";
    private final ProgressRepository progressRepository;
    private final ChannelRepository channelRepository;

    public AskQuestion(ProgressRepository progressRepository, ChannelRepository channelRepository) {
        this.progressRepository = progressRepository;
        this.channelRepository = channelRepository;

        registerTasks();
    }

    @Override
    public boolean supports(Input input) {
        return input.isDirectCommand("ask");
    }

    private void registerTasks() {

        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

        Runnable autoAskQuestionInGroups = () -> {
            // We don't want it to work at night
            if (Time.isOffHours()) {
                return;
            }

            for (Channel channel : channelRepository.list()) {
                ask(channel, IOResolver.resolveFor(channel));
            }
        };

        scheduler.scheduleAtFixedRate(autoAskQuestionInGroups, 0, 2, TimeUnit.HOURS);
    }

    @Override
    public void handle(Input input, Output output) {
        ask(input.channel(), output);
    }

    private void ask(Channel channel, Output output) {
        Optional<ChannelQuestion> result = progressRepository.createRandomForChannel(channel.id);

        if (result.isEmpty()) {
            // TODO share summary. reset progress
            output.write(NO_MORE_QUESTIONS_LEFT);

            return;
        }

        ChannelQuestion newQuestion = result.get();

        Map<String, String> replies = new HashMap<>();
        int optionNum = 0;
        for (String replyVariant : newQuestion.listVariants()) {
            replies.put(replyVariant, String.format("qh answer %s %d", newQuestion.getId().toString(), ++optionNum));
        }

        output.promptChoice(newQuestion.getText(), replies);
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/AskQuestion.kt
package me.nasukhov.tukitalearner.bot

import me.nasukhov.bot.command.Handler
import me.nasukhov.bot.io.Input
import me.nasukhov.bot.io.Output
import me.nasukhov.bot.io.PromptOutput
import me.nasukhov.bot.io.TextOutput
import me.nasukhov.tukitalearner.study.GroupRepository
import me.nasukhov.tukitalearner.study.ProgressTracker
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component("askQuestionCommand")
class AskQuestion(
    private val progressTracker: ProgressTracker,
    private val groupRepository: GroupRepository,
) : Handler {
    companion object {
        private const val NO_MORE_QUESTIONS_LEFT = "У нас пока нет новых вопросов. Проверьте позже"
    }

    override fun supports(input: Input): Boolean = input.isDirectCommand("ask")

    @Transactional
    override fun handle(input: Input): Output {
        val group = groupRepository.findById(input.channel.id)
        val result = progressTracker.createRandomForGroup(group!!)

        if (result.isEmpty) {
            return TextOutput(NO_MORE_QUESTIONS_LEFT)
        }

        val newQuestion = result.get()

        val replies = HashMap<String, String>()
        var optionNum = 0
        for (replyVariant in newQuestion.listVariants()) {
            replies.put(replyVariant, "qh answer ${newQuestion.id} ${++optionNum}")
        }

        return PromptOutput(newQuestion.text, replies)
    }
}
